pinot-controller/src/test/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTestNew.java

```diff
@@ -0,0 +1,1617 @@
+          Map<String, LLCSegmentName> partitionToLatestSegments =
+              segmentManager._partitionAssignmentGenerator.getPartitionToLatestSegments(idealState);
+          LLCSegmentName latestSegment = partitionToLatestSegments.get(String.valueOf(randomlySelectedPartition));
+          idealState = idealStateBuilder.setSegmentState(latestSegment.getSegmentName(), "OFFLINE").build();
+
+          // get old state
+          nPartitions = expectedPartitionAssignment.getNumPartitions();
+          IdealState idealStateCopy = new IdealState(
+              (ZNRecord) znRecordSerializer.deserialize(znRecordSerializer.serialize(idealState.getRecord())));
+          Map<String, Map<String, String>> oldMapFields = idealStateCopy.getRecord().getMapFields();
+
+          if (tooSoonToCorrect) {
+            segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+            // validate nothing changed and try again with disabled
+            Assert.assertEquals(oldMapFields, idealState.getRecord().getMapFields());
+            segmentManager.tooSoonToCorrect = false;
+          }
+
+          segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+
+          verifyRepairs(tableConfig, idealState, expectedPartitionAssignment, segmentManager, oldMapFields);
+        } else {
+          // at least 1 replica of latest segments is in CONSUMING state
+
+          final boolean step2NotDone = random.nextBoolean();
+
+          if (step2NotDone) {
+            // failed after step 1 : hence metadata is DONE but segment is still CONSUMING
+
+            // generate error scenario
+            int randomlySelectedPartition = random.nextInt(nPartitions);
+            Map<String, LLCSegmentName> partitionToLatestSegments =
+                segmentManager._partitionAssignmentGenerator.getPartitionToLatestSegments(idealState);
+            LLCSegmentName latestSegment = partitionToLatestSegments.get(String.valueOf(randomlySelectedPartition));
+            LLCRealtimeSegmentZKMetadata latestMetadata =
+                segmentManager.getRealtimeSegmentZKMetadata(tableName, latestSegment.getSegmentName(), null);
+            segmentManager.updateOldSegmentMetadataZNRecord(tableName, latestSegment,
+                latestMetadata.getStartOffset() + 100);
+
+            // get old state
+            nPartitions = expectedPartitionAssignment.getNumPartitions();
+            IdealState idealStateCopy = new IdealState(
+                (ZNRecord) znRecordSerializer.deserialize(znRecordSerializer.serialize(idealState.getRecord())));
+            Map<String, Map<String, String>> oldMapFields = idealStateCopy.getRecord().getMapFields();
+
+            if (tooSoonToCorrect) {
+              segmentManager.validateLLCSegments(tableConfig, idealState, nPartitions);
+              // validate nothing changed and try again with disabled
+              Assert.assertEquals(oldMapFields, idealState.getRecord().getMapFields());
```

Comment:
I think this assert statement is not doing the comparison as we expect. It just asserts that oldMapFields is a subset of idealState.getRecord().getMapFields()